actors/market/src/state.rs

```diff
@@ -551,20 +572,163 @@ impl State {
     where
         BS: Blockstore,
     {
-        let mut pending_deals = Set::from_root(store, &self.pending_proposals)
-            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to load pending deals")?;
-
-        let rval_pending_deal = pending_deals
+        let mut pending_deals = self.load_pending_deals(store)?;
+        let removed = pending_deals
             .delete(&pending_deal_key.to_bytes())
             .with_context_code(ExitCode::USR_ILLEGAL_STATE, || {
                 format!("failed to delete pending proposal {}", pending_deal_key)
             })?;
 
-        self.pending_proposals = pending_deals
-            .root()
-            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to flush pending deals")?;
+        self.save_pending_deals(&mut pending_deals)?;
+        Ok(removed)
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////
+    // Provider sector/deal operations
+    ////////////////////////////////////////////////////////////////////////////////
+
+    // Stores deal IDs associated with sectors for a provider.
+    // Deal IDs are added to any already stored for the provider and sector.
+    // Returns the root cid of the sector deals map.
+    pub fn put_sector_deal_ids<BS>(
+        &mut self,
+        store: &BS,
+        provider: &Address,
+        sector_deal_ids: &[(SectorNumber, SectorDealIDs)],
+    ) -> Result<(), ActorError>
+    where
+        BS: Blockstore,
+    {
+        let mut provider_sectors = self.load_provider_sectors(store)?;
+        let mut sector_deals = load_provider_sector_deals(store, &provider_sectors, provider)?;
+
+        for (sector_number, deals) in sector_deal_ids {
+            let mut new_deals = deals.clone();
+            let existing_deal_ids = sector_deals
+                .get(sector_number)
+                .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to read sector deals")?;
+            if let Some(existing_deal_ids) = existing_deal_ids {
+                new_deals.deals.extend(existing_deal_ids.deals.iter());
+            }
+            new_deals.deals.sort();
+            new_deals.deals.dedup();
```

Comment:
This is implementing union semantics between the new and existing deals. It's true that I don't expect there to be any duplicates today, but I don't think they should be rejected here. This is dumb state, not application logic. Writing here isn't processing the transition of a deal being activated, it's just storing the final state.

Today, data is activated in a sector only once. When we allow re-snap, it may happen multiple times, including adding new data to a sector while retaining its existing data. Handling that correctly is the responsibility of the actor code that calls this, but I think union a reasonable interface for this state type to offer. The caller can either provide the new final state, or just the delta, whichever is convenient for it.